### DISCUSSION TITLE
Api version to initialize OpenAI Backend

### DISCUSSION THREAD

**Discussion by papayalabs** [AllYourBot/hostedgpt#575] [General]
Now when OpenAI is initialized, there is no option to pass the api_version, so when the api url is passed, for example `https://api.openai.com/`, it automatically becomes `https:// api.openai.com/v1/`.

Perplexity (https://www.perplexity.ai/) has the same OpenAI API protocol (chat.completions) but if we want to use Perplexity Sonar in HostedGPT and put the api url `https://api.perplexity.ai/` will become `https://api.perplexity.ai/v1/` and won't work.

Possible solution: Pass the argument api_version: "" and when using the ChatGPT URL, put v1 already.

**Comment by mattlindsey** [comment]
Did you try it by creating a new API Service in Settings for Perplexity and putting in that URL?  It should use it without tacking on 'v1/`.
I created a new API Service for Ollama and it works fine.

**Reply by papayalabs** [reply]
Yes, I did. But in perplexity it does not work until in code you add api_version: "".

And if I make this change (api_version: "") I would need to update the api-url for ChatGPT to "https://api.openai.com/v1/" and for Ollama "http://localhost:11434/v1/" but Perplexity remains "https://api.perplexity.ai/"

**Reply by mattlindsey** [reply]
Ahh. I see. The ruby-openai gem is adding the 'v1' unless you tell it not to with `api_version: ""`.  

So I think you'll need to either add 'API Version' to the Settings page or something like 'Blank API version' checkbox (checked for Perplexity), which would keep existing records as-is.    Otherwise you need to hard-code some stuff, I think.

Or I suppose always pass `api_version: ""` might be a good solution, as you say.  And do a migration to update existing records.

**Comment by papayalabs** [comment]
Yes @mattlindsey , I think there are this 2 options:

1._ 'Blank API version' checkbox (checked for Perplexity or any other API Service with this configuration )
2._ Put in code api_version: ""  and add  'API Version' to the Settings page. And also do a migration to update existing records.

**Reply by keithschacht** [reply]
Dang, I see what you're talking about with this issue. I didn't realize the gem was modifying the URL. I don't like that it's automatically doing that. Can we disable that behavior? It sounds like doing `api_version: ""` effectively disables it?

If so, then I think we should update our gem initialization code so that we explicitly do `api_version: ""` all of the time, and then we do a migration to modify all OpenAI URLs to change them from `https:// api.openai.com/` to `https:// api.openai.com/v1/`. Then I think the whole system will work as expected — whatever URL is provided is what the API will hit — without us having to make the user think about an additional setting within the UI.

**Reply by mattlindsey** [reply]
I just got Github Models API (https://docs.github.com/en/github-models) working by changing it to `api_version: nil` (instead of `api_version: ""`), and the OpenAI URLs still work unchanged!

@papayalabs If you can confirm it works for Perplexity (and confirm OpenAI too), one of us can do a simple PR without needing a migration.

P.S. Although it fixed Github Models, the `nil` might not work for Perplexity as it seems to still tack on the `v1`, but it might be worth you trying too.

**Reply by papayalabs** [reply]
Hey @mattlindsey , I confirm that Perplexity does not work with` api_version: nil `but it works with `api_version: ""`. It seems that with api_version: nil the gem is still adding the v1 and Perplexity needs the api url without any version. (https://github.com/alexrudall/ruby-openai/blob/0409994e110731d1c1b02a6e10448f4e6d08cb55/lib/openai/client.rb#L24)

I like better the nil than the ""  but in this case it does not work.

**Reply by keithschacht** [reply]
@papayalabs Great, so if i'm understanding correctly would it work to:

1. Change our use of the gem to always do api_version: ""
2. Do a migration to update any existing URLs for openai to add the "v1" directly to the URL

If that plan works, any chance you can tee up a PR?

**Reply by papayalabs** [reply]
Sure @krschacht, but one last thing, why Groq use this URL (https://api.groq.com/openai/v1/) if it will be converted in api.groq.com/open/ai/v1/v1 by the ruby-openai gem ?

(https://github.com/AllYourBot/hostedgpt/blob/b393c4129e40b68474a8527950bbdcf147f3fe68/app/models/api_service.rb#L4C15-L4C46)

Does not make sense that URL. Should be  (https://api.groq.com/openai/) in the current code. I do not have Groq, this works?

**Reply by keithschacht** [reply]
@papayalabs Well, technically openai gem is adding "/v1" to anything using that backend. This means that groq is currently using .../v1/v1 as you pointed out. Oddly, Groq is okay with this.

When you make the change for `api_version: ""` then Groq will stop adding the second v1 so no need to change the URL. BUT when you make this change you will need to do a migration to convert all the openai.com URLs to add a /v1 to them.

Just to be certain of this I quickly tried hitting Groq with a double v1 and a single v1 and it worked in both cases.

**Reply by papayalabs** [reply]
Hey @krschacht ! Very Oddly but if Groq is ok with that great!. Here is the PR --> https://github.com/AllYourBot/hostedgpt/pull/582.

I have tested with ChatGPT and Perplexity and works well. The only strange thing is that Perplexity enter in like a loop when try to do the auto title. Maybe is because this model behave like a search engine. Anyway whatever is the problem is not related to this change.

**Reply by keithschacht** [reply]
@papayalabs thanks for taking this on! I just merged in the PR. Sorry for being slow on reviewing this.
